### PR TITLE
Introduce -m/--module flag to execute a `main` function in a package

### PR DESCRIPTION
**Pull request by KristofferC**
This aims to bring similar functionality to Julia as the `-m` flag for Python which exists to directly run some function in a package and being able to pass arguments to that function.

While in Python, `python -m package args` runs the file `<package>.__main__.py`, the equivalent Julia command (`julia -m Package args`) instead runs `<Package>.main(args)`. The package is assumed to be installed in the environment `julia` is run in.

An example usage could be:

Add the package:
```julia
(@v1.11) pkg> add https://github.com/KristofferC/Rot13.jl
     Cloning git-repo `https://github.com/KristofferC/Rot13.jl`
    Updating git-repo `https://github.com/KristofferC/Rot13.jl`
   Resolving package versions...
    Updating `~/.julia/environments/v1.11/Project.toml`
  [43ef800a] + Rot13 v0.1.0 `https://github.com/KristofferC/Rot13.jl#master`
    Updating `~/.julia/environments/v1.11/Manifest.toml`
  [43ef800a] + Rot13 v0.1.0 `https://github.com/KristofferC/Rot13.jl#master`
```

And then it can be run (since it has a `main` function) via:

```
❯ ./julia/julia -m Rot13 "encrypt this for me" "and this as well"
rapelcg guvf sbe zr
naq guvf nf jryy
```

I'm not sure if `-m/--module` is the best choice but perhaps the association to Python makes it worth it.

### CODE REVIEW

**Comment by Seelengrab** [comment]
Can we make it so that a "less to write" version doesn't clutter the main environment by default? As is, if you want to keep environments clean you have to do something like

```
julia -m Rot13 --project=@Rot13
```

which is super verbose. IMO tying executables/apps/<things-with-`main`> to a seperate environment instead of "whichever Julia happens to run in by default" seems sensibel to me.

**Comment by KristofferC** [comment]
For CLI apps that you would run with just its name (`rot13 args`) I would agree that perhaps having isolated environments for each app would be a good idea. However, something like that would require a lot more infrastructure work. You would probably have a whole Pkg subcommand for that `pkg> app status` etc where you could see all the apps you have installed, update them (`app update rot13`) etc. It would basically be a parallel "universe" from standard packages.

However, for this feature, which is quite a lot more limited in scope, I think it makes sense to basically see it as just a shorthand syntax for loading a package and running a function in it. Having to create and resolve a new environment which might have different versions of what you have installed etc feels like it would be annoying. And then to update things you would have to go activate and update all these environments etc.

**Comment by Seelengrab** [comment]
OK, but isn't this entire PR/idea based around CLI apps? What use other than running CLI apps does this have? We already have `julia -e 'using Foo; Foo.bar()'` for running arbitrary functions after all.

I really don't think citing Python is a good example of how this can be done - python packaging like that is the cause of many problems for sysadmins.

**Comment by KristofferC** [comment]
> We already have julia -e 'using Foo; Foo.bar()' for running arbitrary functions after all.

Yes, but it isn't so easy to propagate arguments to that, for example:

```
~/.julia/packages
❯ ~/julia/julia -m Rot13 `ls`
NofgenpgNytroen
NofgenpgSSGf
NofgenpgYnggvprf
NofgenpgZPZP
NofgenpgGerrf
Npprffbef
Nqncg
NQGlcrf
NqinaprqUZP
```

or work with stdin + piping etc.

**Comment by Seelengrab** [comment]
That's not a good argument for why the "path of least resistance" for CLI apps should clutter the default environment by default.

**Comment by mbauman** [comment]
It seems to me like `julia -m Rot13 "frperg"`  _should_ use the Rot13 from environment that `julia` would use normally — if it's there, anyhow.

Then the question is what happens if Rot13 isn't in that environment.  It'd be an error here.  And perhaps it should automatically do what you suggest (that is, create/use a named env probably with a prompt on creation)... but that seems to be an "extra" feature that's pretty orthogonal, no?

**Comment by Seelengrab** [comment]
> but that seems to be an "extra" feature that's pretty orthogonal, no?

I don't think so - we don't really have CLI apps in the ecosystem precisely because it's a pain to use them with environments (not to mention startup time). Isolating them in an environment by default means we can precompile that environment completely for the `main` invocation, potentially even caching a pkgimage. That's not something that can be done with the ever-changing default environment.

I think it's worthwhile to think the usecase through here, and weigh the up & downsides before settling on an approach. To me, there are very clear UX, update & recompile downsides to making the "default environment" the "easy path", with very little (no?) upside (other than "it's how python does it", which is more of a downside to me..).

**Comment by Seelengrab** [comment]
> or work with stdin + piping etc.

This too is highly dependent on the package in question, not on this interface. I could just as well have written `julia -e 'using Rot13; Rot13.main(ARGS)'` and it'd be exactly the same as what `-m` in this PR is doing; handling `stdin` is no easier in a package with this `-m` approach than one without - after all, this PR does nothing more than pass `ARGS` to a `main` it loaded.

**Comment by KristofferC** [comment]
I think this would mostly be used for packages that can have both functionality as a library and a package. For example, the `Rot13` package could also be used in a non-CLI mode as a standard Julia package. In those cases I think it would be surprising to have e.g. `julia -m Rot13 "frperg"` to use possibly a different version of Rot13 and different version of dependencies compared to `julia -e 'using Rot13; Rot13.reverse_rot13("input")'.

**Comment by KristofferC** [comment]
>  I could just as well have written `julia -e 'using Rot13; Rot13.main(ARGS)'` and it'd be exactly the same as what -m in this PR is doing

Very possible but could you perhaps just show it explicitly to make your point a bit clearer?

**Comment by Seelengrab** [comment]
> In those cases I think it would be surprising to have e.g. julia -m Rot13 "frperg" to use possibly a different version of Rot13 and different version of dependencies compared to `julia -e 'using Rot13; Rot13.reverse_rot13("input")'.

That sounds much more like an argument for "don't clutter the default environment with CLI apps because it will be held back by other, unrelated stuff in that environment" than an argument for making this use the default environment by default.

Do we really want to have lots of bug reports about old versions of CLIs floating around in the ecosystem? I think the SciML ecosystem still struggles with those because of old docs pages & old tutorials showing up on google.

> Very possible but could you perhaps just show it explicitly to make your point a bit clearer?

Sure:

```
[sukera@tower ~]$ julia -q --project=@Rot13 -e 'import Pkg; Pkg.add("https://github.com/KristofferC/Rot13.jl")
[sukera@tower ~]$ julia -q --project=@Rot13 -e 'using Rot13; Rot13.main(ARGS)' "encrypt this for me" "and this as well"
rapelcg guvf sbe zr
naq guvf nf jryy
```

I've put the package into its own environment because I don't want random CLIs to clutter my main environment, and I don't want those to be held back by other random stuff in my main environment.

**Comment by mbauman** [comment]
The key question is what should `julia -m Rot13 "frperg"` do if Rot13 exists in the environment that `julia` uses.  I agree with Kristoffer — I think it'd be very confusing if it'd come from a different place.

**Comment by Seelengrab** [comment]
I don't think it should. There should be a very clear seperation between uses as a library and uses as a packaged CLI. I think this should always use its bespoke environment.

If we want to look for other ecosystems for inspiration, can we take a look at e.g. Rust? Their CLI game is well known to be amazing, so why not take a page out of their book?

**Comment by KristofferC** [comment]
> Sure:

Thanks!

So to be clear, the concrete suggestion here is for `julia -m Foo` to default to `julia --project=@Foo -m Foo`? Some follow up questions to that:

- Wouldn't it be annoying having to activate and add the package to that environment? Or do you want a `pkg> app install Foo` that creates some "internal" environment (e.g. `.julia/environments/apps/Foo_$FOO_UUID`) and resolves in there and then `julia -m` activates that environment?
- If the environment is internal, how do you check what dependencies are installed / update them etc? `pkg> app status Foo`, `pkg> app update Foo`?
- Are the dependencies of Foo "locked in" when the package is added or does Foo itself control that with some explicitly checked in manifest?

**Comment by adienes** [comment]
if I understand correctly
```
julia --project=@Rot13 -e 'import Pkg; Pkg.add("Rot13")'
```

this only happens _once_, on install?

I'm no expert on CLI design, but imo makes more sense for `julia -m Rot13` to default use the same environment as `julia`

However, definitely also feels like a reasonable ask to make isolated envs for cli apps ergonomic, but I think that's orthogonal feature and shouldn't be default?

**Comment by KristofferC** [comment]
As I said before, I do think there is a difference between having a "driver" like `julia -m Rot13 secret` vs something that looks like a standalone executable (`rot13 secret`). In the first case you really know you are doing something julia specific and it is natural to think that standard julia behavior applies to it. In the second, "anything goes" since in many cases you might not even know you end up executing julia code so then it would be surprising that e.g. changes to the default julia environment would influence how the standalone looking `rot13` behaves.

**Comment by fredrikekre** [comment]
> As I said before, I do think there is a difference between having a "driver" like julia -m Rot13 secret vs something that looks like a standalone executable (rot13 secret). In the first case you really know you are doing something julia specific and it is natural to think that standard julia behavior applies to it. In the second, "anything goes" since in many cases you might not even know you end up executing julia code so then it would be surprising that e.g. changes to the default julia environment would influence how the standalone looking rot13 behaves.

I agree fully, and I had typed up an almost identical comment but got sniped. In particular, the current behavior lets me choose/control the environment, which can be useful sometimes. That wouldn't be possible with the other suggested approach.

Also, people should really relax about the global environment, it is amazingly convenient. I rarely, if ever, run into problems related to this. Maybe I just don't use "troublesome" ecosystems or combinations of packages.

image

(Disclaimer: this is just a meme documenting my personal development with environments in Julia over the years and not meant to be offensive to anyone :slightly_smiling_face: )

**Comment by mbauman** [comment]
I see this as being as much about ~~`python -m http.server 8080`~~ `julia -m HTTP serve 8080` as it is about `julia -m DataAnalysis.Preprocess data/*.csv` or `julia -m DataAnalysis.Dashboard`. In the latter cases, I'd expect to be actively working on an active project — and could see this as being a useful paradigm for accessing _lots_ of entry points therein.  It'd be hair-pullingly painful if `-m` silently checked out a different version from what I'm actively doing!

As far as the pains of add/activate/instantiate and (potentially) dedicated environments, those span many more usages than just `-m Module`.  E.g., https://github.com/JuliaLang/Pkg.jl/issues/1415 and https://github.com/JuliaLang/Pkg.jl/issues/2401.  And I do think that `-m` do something useful in the package-not-found-error case that helps smooth over some of these pain points.

**Comment by PallHaraldsson** [comment]
How about on first use:

```
julia -m Rot13 "encrypt this for me" "and this as well"
```

would make an empty environment named Rot13, download Rot13 for you, the latest version, and on subsequent use, use that environment (and people could customize it in the meantime)? There's a question where it would be stored, in the current dir or elsewhere.

If Rot13 exists in the global environment (and maybe in each case if there's a conflict, then it could ask, "Do you want to use the global environment (once)? Or copy from it into a local environment? Or download latest version?".

Could a local environment point to the global environment somehow? Or can a local environment program already do that selectively, if it wants?

It seems you might want to support this the same way (at least now since not registered):
```
julia -m https://github.com/KristofferC/Rot13.jl "encrypt this for me" "and this as well"
```

>It seems to me like julia -m Rot13 "frperg" should use the Rot13 from environment that julia would use normally — if it's there, anyhow.

If you feel like that, and people's ideas differ, should it maybe work like that with -M Rot13?

>people should really relax about the global environment, it is amazingly convenient. I rarely, if ever, run into problems related to this.

I run into trouble all the time, I can't install packages, since my environment is too cluttered. FYI: Python has Zip-apps, that I think are though not much used, that allows a whole programs with all dependencies stored in one zipped file with .pyz ending. Julia wouldn't needs such if you can just point to an URL.

**Comment by Seelengrab** [comment]
> So to be clear, the concrete suggestion here is for julia -m Foo to default to julia --project=@Foo -m Foo? Some follow up questions to that:

All of these issues literally only exist because you've made the `julia` binary do all that work. The `julia` binary already has so many overloaded uses that there is hardly any design space left to do a better job - adding even more kitchen sinks to this as with this PR won't help with that.

Taking a page out of the rust book, why must `julia` be the driver binary for this?

> Wouldn't it be annoying having to activate and add the package to that environment? Or do you want a pkg> app install Foo that creates some "internal" environment (e.g. .julia/environments/apps/Foo_$FOO_UUID) and resolves in there and then julia -m activates that environment?

The latter, similar to [`cargo install`](https://doc.rust-lang.org/cargo/commands/cargo-install.html).

Though again, there is no need to have `julia` then also be the driver to run this. There is design opportunity here to actually make CLIs nice to use without having to type all of `julia -m Foobar`. If something is to be used as a CLI, it ought to be as standalone invokable as possible.

> If the environment is internal, how do you check what dependencies are installed / update them etc? pkg> app status Foo, pkg> app update Foo?

It's a regular environment, so you can do something like `]activate app Foobar` to activate the app/CLI environment and then do all the regular Pkg operations to your hearts content. It's a regular Pkg environment after all.

> Are the dependencies of Foo "locked in" when the package is added or does Foo itself control that with some explicitly checked in manifest?

The dependencies of the CLI resolve as any other package environment does. If `Foo` compat bounds dependencies, all compat bounds apply as usual. The CLI can be updated as with any other environment, by activating it and calling `]up`.

> If I want to actually use a package in my environment would I do julia --project=@v1.11 Foo and make sure to match that project env with the julia version?

No, you can use `julia --project=@.`, just as you can with any other environment. This defaults to the global environment when not in a project directory, and to a project environment when in a project directory. Since environments stack, a CLI package installed in your global environment will still be accessible, even if a project environment is activated.

Though again, that is _only_ a limitation of being forced to use `julia` as the driver binary in the first place. This issue doesn't exist if you use a different driver binary (like rust does) for the compiler vs. CLI/Pkg management. 

> (Disclaimer: this is just a meme that documents my personal development with environments in Julia over the years and not meant to be offensive to anyone 🙂 )

I find this sort of commentary unnecessarily provocative, counterproductive and generally inappropriate. If you don't have anything productive to add to the discussion, please keep your memes to yourself. You're painting any usecase other than your own as being whiny and annoying - perhaps consider that there are different folks with different requirements who have good reasons for wanting to make the path of least resistance more welcoming to traditional software folk.

---

I find it generally discouraging that every single time I try to advocate for learning from the mistakes of other languages/ecosystems, people seem to double down on reinventing the wheel.

**Comment by KristofferC** [comment]
> Though again, there is no need to have julia then also be the driver to run this. 

So it seems you are arguing for having (perhaps only) the "standalone" app functionality I described in https://github.com/JuliaLang/julia/pull/52103#issuecomment-1804771847 and  https://github.com/JuliaLang/julia/pull/52103#issuecomment-1804671551, (which is more similar to https://python-poetry.org/docs/pyproject/#scripts and https://doc.rust-lang.org/cargo/reference/cargo-targets.html#binaries.). 

 I agree that that functionality is useful and it is also something I am actively thinking about (see e.g a screenshot from some thoughts scratch document about this):

image

Maybe your worry is that CLI support in Julia would end with this PR in which case I understand (and would share) your worry. However, I would only consider this PR as a "little brother" to the standalone executable functionality not a replacement. This is still useful in smaller scale but you would indeed not want to write the next Dropbox using it (`julia -m Dropbox` lol).

As I wrote in https://github.com/JuliaLang/julia/pull/52103#issuecomment-1804771847, there is a lot of design for the standalone CLI support and I would be very grateful for help with the design there. If you write to me I'll invite you to my scratch design document I screenshotted from above and we can iterate there.

**Comment by KristofferC** [comment]
>  Since environments stack, a CLI package installed in your global environment will still be accessible, even if a project environment is activated.

While true, it will look up dependencies in your project environment first which may be incompatible with other dependencies that gets found from the global env so it can cause quite weird bugs. I wouldn't want any standard use case to rely on it.

> I find it generally discouraging that every single time I try to advocate for learning from the mistakes of other languages/ecosystems, people seem to double down on reinventing the wheel.

Keep fighting the good fight, your input is valuable and appreciated :).

**Comment by Seelengrab** [comment]
> Maybe your worry is that CLI support in Julia would end with this PR in which case I understand (and would share) your worry. However, I would only consider this PR as a "little brother" to the standalone executable functionality not a replacement. This is still useful in smaller scale but you would indeed not want to write the next Dropbox using it (julia -m Dropbox lol).

Yes indeed - having this be the "main" thing for apps is part of it, but another is that having multiple ways to do kind-of similar-ish things (run a CLI) with potentially different behaviors is confusing to say the least (as shown in this thread). Thus, I think we either have to reconcile the ideas somehow, or decide on a canonical way this ought to work.

> I would be very grateful for help with the design there. If you write to me I'll invite you to my scratch design document I screenshotted from above and we can iterate there.

Sounds good - I'll write you on Slack.

File: base/client.jl [hunk 1]
```diff
@@ -292,6 +292,30 @@ function exec_options(opts)
         elseif cmd == 'E'
             invokelatest(show, Core.eval(Main, parse_input_line(arg)))
             println()
+        elseif cmd == 'm'
+            mods = split(arg, ".")
+            pkg = popfirst!(mods)
+            # Some more validation of `mods` and `pkg`?
+            m = Base.require(Main, Symbol(pkg))
```
Comment (Keno) [inline base/client.jl:299]: So, technically this is all that is required if the module exports its main function - that was one of the explicit design goals of `@main` that you could use it as `-e 'using Foo'` to get the python app like behavior. I think it's fine to further shortcut that to `-m`, but I'm not sure about duplicating the `main` invocation logic here.

Comment (KristofferC) [reply base/client.jl:299]: Okay I agree with that. So if an app want to have multiple "functionalities" (like the proposed `julia -m Foo/Bar` to run `Foo.Bar.main`) would you do that as `julia -m Foo --bar` or still do `julia -m Foo/Bar` and then have that still call `Foo.main` and look at something like `Base.PROGRAM_FILE` to determine what to execute.

**Comment by lmiq** [comment]
The issue I see with using the global environment is that such a call will be hardly reproducible. Someone incorporating a `julia -m Foo arg` into a script will have the script broken whenever changes to the global environment occur. 

Of course then one can manage a specific environment for that manually, but then the utility of the default behavior is just limited to saving a few keystrokes, and hardly recomendable to a general user.

Since code cashing, using clean temporary (or not) environments became convenient (locally available packages install quickly), and IMHO Julia should move in the direction to encourage keeping the global environment as clean as possible. After you install CUDA, Makie, or another heavyweight precompiled package, getting random precompilations is really annoying. I bet no one wants to run `julia -m Foo args` and suddenly wait for five minutes.

**Comment by KristofferC** [comment]
@Keno, can you look at this new implementation?

**Comment by KristofferC** [comment]
> The issue I see with using the global environment is that such a call will be hardly reproducible. Someone incorporating a julia -m Foo arg into a script may have the script broken whenever changes to the global environment occur.

Having these be reproducible is not the goal of this specific feature though. It is mostly a convenience to be able to write `julia -m IJulia` or `julia -m Pluto`.

For "full-scale apps" something more powerful is probably needed yes but that would look completely different (https://github.com/JuliaLang/julia/pull/52103#issuecomment-1804771847).

**Comment by Seelengrab** [comment]
> Having these be reproducible is not the goal of this specific feature though. 

I think everyone against `-m` is well aware that this is not a "goal" of your PR. We're just opposed to having this NOT be a goal, since designing only with `Pluto` or `IJulia` in mind means that other usecases will suffer disproportionally more than having `Pluto` or `IJulia` be in its own environment by default.

**Comment by KristofferC** [comment]
>  We're just opposed to having this NOT be a goal, since designing only with Pluto or IJulia in mind means that other usecases will suffer disproportionally more

There shouldn't be any extra suffering here :). Packages that want to expose this functionality to users would implement a `@main` method which is what they would do with the more "advanced scheme" as well. And for this advanced scheme, we wouldn't want to use a julia driver anyway so this new command line argument also does not occupy any design space there. And this feature is invoked explicitly by a user so anyone who does not like this can just choose not to use it and instead manually manage the environments and explicitly call `IJulia.main()` just like how it is done now. So I don't see any suffering from either users or package developers by having this option being available. But I see quite a bit of convenience from it in various scenarios.

**Comment by Seelengrab** [comment]
> There shouldn't be any extra suffering here :). 

The extra suffering *is* the clobbering of the default environment with (potentially incompatible) packages.

I don't think anyone answered my question about how such held back packages should be dealt with:

> On the other hand, how do you propose to deal with held back packages because of CLI apps? Or the other way around?

**Comment by KristofferC** [comment]
> The extra suffering is the clobbering of the default environment with (potentially incompatible) packages.

Pkg will ensure they are compatible but installing many packages in the same environment has the same drawback as it currently has. Although I should note that in some cases, e.g. IJulia, you should probably in fact have it executed in the same environment as the code you are running. IJulia loads packages into your running Julia session and if it runs from a separate environment from your code it is possible the IJulia dependencies are incompatible with what the packages in your code environment is resolved for.

> On the other hand, how do you propose to deal with held back packages because of CLI apps? Or the other way around?

That is unchanged from the status quo, you can for example use `--project=@xxx`. But again, this PR doesn't attempt to have a solution for that. It is indeed possible that you wouldn't want to use `julia -m` to start e.g. an e-book reader that you wrote in Julia. For that e.g. https://github.com/Roger-luo/Ion might be better.

**Comment by Seelengrab** [comment]
> IJulia loads packages into your running Julia session and if it runs from a separate environment from your code it is possible the IJulia dependencies are incompatible with what the packages in your code environment is resolved for.

That sounds like an additional argument for not dumping it into the default environment, precisely to not have this issue.

> It is indeed possible that you wouldn't want to use julia -m to start e.g. an e-book reader that you wrote in Julia

How/why is Pluto or IJulia qualitatively different here? It's not necessary to have Pluto installed in the environment of the notebook either, for example; Pluto already manages the environment of the notebook for you.

**Comment by KristofferC** [comment]
> That sounds like an additional argument for not dumping it into the default environment, precisely to not have this issue.

Yeah, using an environment with IJulia together with the code you want to run is likely the best idea there.

**Comment by jakobnissen** [comment]
I'm not convinced this provide any real benefit. If I understand correctly, `julia -m MyPackage` is equivalent to `julia -e 'using MyPackage: main; main()'`. Alternatively, I could just make a two-line script that loads the package and executes it. I don't see any real value in a few feature just to save that handful of keystrokes.

Indeed, this PR may do more harm than good, because it seems to me to solve a non-problem, while not addressing the current problems with executing Julia code. I believe the more pressing gotchas with running Julia scripts from command-line are:
* Having to create a global environment and run the code in that environment
* Disabling stacked environments to make sure dependencies from higher environment aren't relied on when testing the script
* Disabling `startup.jl` which may interfere with the loaded environment
* The various issues stemming from our JIT model, including latency and the absence of a single executable file (addressing these is way outside the scope of this PR, I know)

This is not to disrepect the work you're doing, Kristoffer, to make Julia more easy to execute. But as this PR doesn't address any of the issues above, I worry that it may in effect add a feature which I would never recommend anyone to use, and which will clutter the "solution space" and make it harder to navigate for Julia newbies.

For comparison, Comonicon.jl creates an executable Julia file with the following content, which does address all my concerns above:
```julia
#!/usr/bin/env bash
#=
JULIA_PROJECT=/home/jakob/.julia/scratchspaces/b91f9d24-e2a8-4b03-b25e-993ad86b9958/env \
exec /home/jakob/.julia/juliaup/julia-1.10.0-beta3+0.x64.linux.gnu/bin/julia \
    --startup-file=no \
    --color=yes \
    --compile=yes \
    --optimize=2 \
    -- "${BASH_SOURCE[0]}" "$@"
=#

# generated by Comonicon for the CLI Application binbench
using MyPackage
exit(MyPackage.command_main())
```

Perhaps the `-m` flag in this PR could include the above changes (and thus this PR would also close #42205)?

**Comment by KristofferC** [comment]
> If I understand correctly, julia -m MyPackage is equivalent to julia -e 'using MyPackage: main; main()'

It is not equivalent, with `-m` you know the intention is to execute the entry point of an installed package and can e.g. give better error messages (as this PR already takes advantage of) while with `-e` you Julia is in "evaluate arbitrary expression"-mode. And somewhat ironically you already had a mistake in what you wrote to be the equivalent expression.

> I believe the more pressing gotchas with running Julia scripts from command-line are:

This PR doesn't really talk about running Julia scripts from the command line.

> Perhaps the -m flag in this PR could include the above changes

https://github.com/JuliaLang/julia/issues/42205 does not seem to have much to do with this. #42205 is an issue because you are running the code in Main while you are here running it inside a package.

**Comment by KristofferC** [comment]
I've been toying a bit with the "standalone app" stuff I've mentioned before and I realize that in the implementation of that this would still be useful to have. The reason for that is that even in that case in the end you want to actually start Julia and run the entry point to the app.

You can either use `-e` which is inelegant and inefficient since it requires evaluating arbitrary code (and thereby requiring having an interpreter or compiler at hand). Comonicon.jl effectively uses `-e` since it runs a non-precompiled file with julia code. Or you can use `-m` (together with suitable load paths etc set up) which does exactly what you want and have the code to execute it available to run in the sysimage.

I think a command line like this is a very natural evolution to the `@main`-concept.

**Comment by KristofferC** [comment]
Bump... Does the last comment manage to persuade any of the nay-sayers? If not, what is the alternative? If the answer is `-e` it seems bad to have to add extra latency and force the existence of a julia compiler to run the arbitrary code in `-e`, as well as removing the possibility for better error messages since we do not know that a main entry point is being executed.

**Comment by Seelengrab** [comment]
From my POV, the concern with dumping everything in the default environment (as this PR currently does) has not been addressed :shrug: I'm aware that in a scenario of a standalone app with `LOAD_PATH` set up correctly etc. this is not a concern, but in naive `julia -m Foo` usage, it still very much is.

**Comment by KristofferC** [comment]
> From my POV, the concern with dumping everything in the default environment (as this PR currently does) has not been addressed

Here is the PR that builds on top of this https://github.com/JuliaLang/Pkg.jl/pull/3772

**Comment by MasonProtter** [comment]
I'm not one of the naysayers, but I'd just like to lend my voice in favour of this implementation and choice of environment semantic. 

This is the most natural and straightforward extension of the way that the `@main` entrypoint currently works: simply extending it to allow for a different package than `Main` to provide their entrypoint. Having this silently and automatically create a separate environment seems bad.

**Review by vchuravy — APPROVED** [review]

**Comment by Seelengrab** [comment]
As I said above, if this were exclusive to Apps, I wouldn't have a problem with this. However, the overall commit message and NEWS entry don't make a mention of that, instead likening this to what Python does with having everything that's installed being dumped into the same "environment".

It's this standalone usage, separate from the usage as Apps, that the "naysayers" are worried about. In order to use this feature standalone, the most straightforward way would be to just add everything someone wants to use into the main environment, with the consequences outlined above. This has not been addressed.

Moreover, the current design in https://github.com/JuliaLang/Pkg.jl/pull/3772 simply sets the JULIA_LOAD_PATH to exclusively use the installed app environment. To my understanding, the same thing could be achieved through an empty `JULIA_LOAD_PATH`, combined with passing the environment to be used through `--project`. Requiring the use of `--project` when using `-m` would completely alleviate my concerns, since that then creates an incentive for new users to go through our App system, instead of just thinking "oh it's like python" and messing up their main environment with potentially outdated packages due to version downgrades.

**Comment by Seelengrab** [comment]
As an alternative - can the flag just be made much longer to disincentivize its use? After all, if its intended for usage from Apps, there should be no harm in having a longer flag.

**Comment by Seelengrab** [comment]
As an alternative - can the flag just be made much longer to disincentivize its use? After all, if its intended for usage from Apps (& thus autogenerated shims), there should be no harm in having a longer flag. @lmiq since you also mentioned opposition to the current behavior, would that be a good compromise for you?

**Comment by adienes** [comment]
I like the current state of the PR and am excited to use `-m` ; I certainly don't think its use should be disincentivized

if you want to control the project you're running a module in, `--project` is always right at your fingertips

**Comment by Seelengrab** [comment]
> I like the current state of the PR and am excited to use -m ; I certainly don't think its use should be disincentivized

Ok, how do you propose to deal with held back packages because of liberal usage of standalone `-m`? Or the associated long compile times when that main environment is modified and now all your CLIs have to recompile due to invalidation?

As far as I've noticed, that hasn't been addressed by anyone in favor of liberal usage of standalone `-m` here. That's all I'm inquiring/trying to solve here.

**Comment by adienes** [comment]
> how do you propose to deal with held back packages because of liberal usage of standalone -m?

this does not seem like a likely problem to me so I don't think it needs a solution

> Or the associated long compile times when that main environment is modified and now all your CLIs have to recompile due to invalidation?

if it becomes an issue, I'll put each cli into its own environment. but I don't want that to happen by default

**Comment by MasonProtter** [comment]
It's no worse than regular scripts invoked as `julia foo.jl`, and actually a little better since these CLIs can be tracked by the global environment and thus will do their re-compilation when you do `]up`.

**Comment by KristofferC** [comment]
I'll wait with merging this until after branching for 1.11. Looking at the comments here I think there is more or less a consensus to not do some behind the scene project modification.

**Pull request merged**